webui/react/src/stores/polling.ts

```diff
@@ -0,0 +1,110 @@
+import handleError from 'utils/error';
+
+const POLLING_DELAY_MS = 5000;
+const RETRY_DELAY_MS = 1000;
+const MAX_RETRY_DELAY_MS = 60_000;
+
+// Less than 0 means to keep trying indefinitely.
+const MAX_RETRY = -1;
+const EMPTY_FUNCTION = () => {
+  return;
+};
+
+interface PollingOptions {
+  args?: unknown[];
+  condition?: boolean;
+  delay?: number;
+  maxRetry?: number;
+}
+
+/**
+ * To use this polling store...
+ *
+ * 1. Take your store and extend this `PollingStore` class.
+ *
+ *    class MyStore extends PollingStore { ... }
+ *
+ * 2. Then you simply need to override the `poll` function.
+ *    This is executed during each polling iteration.
+ *    You can also override the `pollCatch` function, but it is completely optional.
+ *    `pollCatch` is executed when an error is encountered during the `poll`
+ *    such as network issues or API failures.
+ *
+ * 3. Call the `startPolling` and `stopPolling` functions to
+ *    start and stop the polling. The most common way is to do this
+ *    in a `React.useEffect` as such:
+ *
+ *    useEffect(() => myStore.startPolling(), []);
+ *
+ *    `startPolling` returns `stopPolling` appropriately so during a
+ *    React unmount it will call the `stopPolling` routine automatically.
+ */
+abstract class PollingStore {
+  protected canceler?: AbortController;
+  protected pollingTimer?: NodeJS.Timer;
+  protected pollingArgs: unknown[] = [];
+  protected pollingDelay = POLLING_DELAY_MS;
+  protected pollingRetry = 0;
+  protected maxRetry = MAX_RETRY;
```

Comment:
question: these shouldn't these `protected` be `private (#)`?

Reply:
`canceler` in particular has to be `protected` because the extended class needs to access it often when defining custom `poll` function. The other ones may be useful to access also in the extended class but as of now they are not being used other than privately.

Reply:
how about make all class members private and make getter / setter protected? i think its good for encapsulation

Reply:
the reason is the same as [this comment](https://github.com/determined-ai/determined/pull/6326#discussion_r1156275446).
we can change the variables accidentally with the current approach. getter / setter can avoid the issue.

Reply:
if `canceler` is defined as:
```
protected readonly canceler = new AbortController();
```
You should not be able to modify it. So specifically:
```
class MyStore extends PollingStore {
  test() {
    this.canceler = new AbortController();
  }
}
```
You'll see a squiggly line under `canceler` with `Cannot assign to 'canceler' because it is a read-only property`

Reply:
i meant ts `readonly` works on only compile time, so if we want some safety for runtime, getter / setter would be useful. If its not the case, im fine with the current way too.